src/track/serato/beatgrid.cpp

```diff
@@ -0,0 +1,401 @@
+                              << pNonTerminalMarker->positionSecs()
+                              << "< 0";
+            return false;
+        }
+
+        if (pNonTerminalMarker->positionSecs() <= previousBeatPositionMillis) {
+            kLogger.warning() << "Parsing SeratoBeatGrid failed:"
+                              << "Non-terminal marker's position"
+                              << pNonTerminalMarker->positionSecs()
+                              << "must be greater than the previous marker's position"
+                              << previousBeatPositionMillis;
+            return false;
+        }
+        previousBeatPositionMillis = pNonTerminalMarker->positionSecs();
+
+        nonTerminalMarkers.append(pNonTerminalMarker);
+    }
+
+    // Read last (terminal) beatgrid marker
+    if (stream.readRawData(buffer, sizeof(buffer)) != sizeof(buffer)) {
+        kLogger.warning() << "Parsing SeratoBeatGrid failed:"
+                          << "unable to read terminal marker data";
+        return false;
+    }
+
+    QByteArray markerData = QByteArray(buffer, kMarkerSizeID3);
+    SeratoBeatGridTerminalMarkerPointer pTerminalMarker =
+            SeratoBeatGridTerminalMarker::parseID3(markerData);
+    if (!pTerminalMarker) {
+        kLogger.warning() << "Parsing SeratoBeatGrid failed:"
+                          << "Unable to parse terminal marker!";
+        return false;
+    }
+
+    if (pTerminalMarker->bpm() <= 0) {
+        kLogger.warning() << "Parsing SeratoBeatGrid failed:"
+                          << "Terminal marker's BPM"
+                          << pTerminalMarker->bpm()
+                          << "must be greater than 0";
+        return false;
+    }
+
+    if (pTerminalMarker->positionSecs() < 0) {
+        kLogger.warning() << "Parsing SeratoBeatGrid failed:"
+                          << "Non-terminal marker has invalid position"
+                          << pTerminalMarker->positionSecs()
+                          << "< 0";
+        return false;
+    }
+
+    if (pTerminalMarker->positionSecs() <= previousBeatPositionMillis) {
+        kLogger.warning() << "Parsing SeratoBeatGrid failed:"
+                          << "Terminal marker's position"
+                          << pTerminalMarker->positionSecs()
+                          << "must be greater than the previous marker's position"
+                          << previousBeatPositionMillis;
+        return false;
+    }
+
+    // Read footer
+    //
+    // FIXME: This byte has caused some headache because I have not the
+    // slightest idea what this value could be. Apparently it's random, because
+    // it changes even when entering Serato's "Edit Grid" mode and then leaving
+    // it immediately without making any changes.
+    // For now, we only read it to be able to dump the exact same byte sequence
+    // later on.
+    quint8 footer;
+    stream >> footer;
```

Comment:
I can imagine Serato uses it as "dirty" flag. I am afraid we need to toggle it as well in case we alter the beat grid in Mixxx. Can you verify it?

Reply:
I can check if it new values are only picked up by Serato if that value is different (if that's what you mean). However, Serato is also doing strange stuff elsewhere (e.g. they sometimes append a random byte for base64 encoded flac metadata). My guess was that they have a messed up length check and write a  byte of uninitialized memory :D

Reply:
That seems also be likely, in particularly if it is not constantly increasing.

Reply:
It's not increasing.